File: packages/shared-types/typescript/src/correlation.ts
/**
 * correlation_id helpers (TypeScript).
 *
 * Generation rules (priority order):
 *   1. Natural ID if one exists (PR number, Slack thread ID).
 *   2. hash(source + object_id + time_window) + uuid suffix.
 *   3. API assigns a fallback if the producer omits it.
 */

function timeWindowBucket(when: Date, windowSeconds = 60): number {
  return Math.floor(when.getTime() / 1000 / windowSeconds);
}

async function sha256Hex(data: string): Promise<string> {
  const enc = new TextEncoder().encode(data);
  const digest = await crypto.subtle.digest("SHA-256", enc);
  return Array.from(new Uint8Array(digest))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

function uuid4Suffix(): string {
  return crypto.randomUUID().replace(/-/g, "").slice(0, 8);
}

export async function deriveCorrelationId(opts: {
  source: string;
  objectId: string;
  naturalId?: string | null;
  when?: Date;
  windowSeconds?: number;
}): Promise<string> {
  const { source, objectId, naturalId, when = new Date(), windowSeconds = 60 } = opts;

  if (naturalId) return `${source}:${naturalId}`;

  const bucket = timeWindowBucket(when, windowSeconds);
  const digest = (await sha256Hex(`${source}|${objectId}|${bucket}`)).slice(0, 16);
  return `${source}:${digest}:${uuid4Suffix()}`;
}

File: packages/shared-types/python/src/mycelium_shared_types/correlation.py
"""correlation_id helpers.

Generation rules (priority order):
  1. Natural ID if one exists (PR number, Slack thread ID).
  2. hash(source + object_id + time_window) + uuid suffix to prevent burst collisions.
  3. API assigns a fallback if the producer omits it.

Use ``derive_correlation_id`` for rules (1) and (2). Rule (3) is implemented in apps/api.
"""

from __future__ import annotations

import hashlib
import uuid
from datetime import datetime, timezone
from typing import Optional


def _time_window_bucket(when: datetime, window_seconds: int = 60) -> int:
    """Bucket a timestamp into a fixed-size window. Defaults to 60-second buckets."""
    return int(when.timestamp()) // window_seconds


def derive_correlation_id(
    *,
    source: str,
    object_id: str,
    natural_id: Optional[str] = None,
    when: Optional[datetime] = None,
    window_seconds: int = 60,
) -> str:
    """Derive a correlation_id following the priority rules above.

    If a ``natural_id`` is provided we use it directly (rule 1). Otherwise we hash
    ``source + object_id + time_window_bucket`` and append a uuid4 suffix to prevent
    collisions between concurrent producers in the same window (rule 2).
    """
    if natural_id:
        return f"{source}:{natural_id}"

    when = when or datetime.now(timezone.utc)
    bucket = _time_window_bucket(when, window_seconds)
    digest = hashlib.sha256(
        f"{source}|{object_id}|{bucket}".encode("utf-8")
    ).hexdigest()[:16]
    suffix = uuid.uuid4().hex[:8]
    return f"{source}:{digest}:{suffix}"
